src/herder/TxSetFrame.cpp

```diff
@@ -33,43 +34,34 @@ namespace stellar
 
 using namespace std;
 
-TxSetFrame::TxSetFrame(Hash const& previousLedgerHash)
-    : mHash(std::nullopt)
-    , mValid(std::nullopt)
-    , mPreviousLedgerHash(previousLedgerHash)
-{
-}
-
-TxSetFrame::TxSetFrame(Hash const& networkID, TransactionSet const& xdrSet)
-    : mHash(std::nullopt), mValid(std::nullopt)
+// Construct a valid TxSetFrame from a list of transactions (e.g. from the
+// TransactionQueue). We make sure the transactions are sorted in (and
+// maintains) hash order and computes the contents hash on contruction.
+TxSetFrame::TxSetFrame(Hash const& previousLedgerHash,
+                       Transactions const& transactions)
+    : mPreviousLedgerHash(previousLedgerHash)
+    , mTxs(TxSetUtils::sortTxsInHashOrder(transactions))
+    , mTxsInValidHashOrder(true)
+    , mHash(TxSetUtils::computeContentsHash(previousLedgerHash, mTxs))
 {
-    ZoneScoped;
-    for (auto const& env : xdrSet.txs)
-    {
-        auto tx = TransactionFrameBase::makeTransactionFromWire(networkID, env);
-        mTransactions.push_back(tx);
-    }
-    mPreviousLedgerHash = xdrSet.previousLedgerHash;
 }
 
-static bool
-HashTxSorter(TransactionFrameBasePtr const& tx1,
-             TransactionFrameBasePtr const& tx2)
+TxSetFrame::TxSetFrame(Hash const& previousLedgerHash)
+    : TxSetFrame(previousLedgerHash, Transactions{})
 {
-    // need to use the hash of whole tx here since multiple txs could have
-    // the same Contents
-    return tx1->getFullHash() < tx2->getFullHash();
 }
 
-// order the txset correctly
-// must take into account multiple tx from same account
-void
-TxSetFrame::sortForHash()
+// Contruct a TxSetFrame from an xdrSet, which may directly come from the
+// overlay and be invalid (in the incorrect order), in which case we mark
+// this set invalid and do not comptue its hash.
```

Comment:
As I've mentioned in another comment, I would like to decouple XDR validation from construction; I don't think that validation in constructor is a good approach. I may take care of that in my PR though.